File: ropchain/solver.py
import ropchain
from gadgets import gadget, setVal, toZero, asm, util
import arch
from struct import pack
import itertools
import copy
import re


def _solve(dests, gadgets, base, cond, proc):
    if arch.arch == arch.X86:
        regs = {'eax', 'ebx', 'ecx', 'edx', 'edi', 'esi', 'ebp'}
    elif arch.arch == arch.AMD64:
        regs = {'rax', 'rbx', 'rcx', 'rdx', 'rdi', 'rsi', 'rbp',
                'r8', 'r9', 'r10', 'r11', 'r12', 'r13', 'r14', 'r15'}
    ropChains = {reg: find(reg, dests[reg], gadgets, set(), cond, proc) for reg in dests}
    ropChains = dict((k, v) for k, v in ropChains.iteritems() if v)
    solvable = set(ropChains.keys())

    remains = set(dests.keys()) - solvable
    ans = None
    isFail = True

    for rs in itertools.permutations(remains):
        tmpAns = ropchain.ROPChain(None)
        canUse = copy.deepcopy(regs)
        for reg in rs:
            canUse.remove(reg)
            tmp = find(reg, dests[reg], gadgets, canUse, cond, proc)
            if tmp is None:
                tmpAns = None
                break
            tmpAns = tmpAns + tmp if tmpAns else tmp

        if tmpAns is None:
            continue
        isFail = False
        ans = util.optMin(ans, tmpAns)

    if isFail:
        return None
    r = sum([ropChains[reg] for reg in ropChains], ans)
    r.setBase(base)
    return r

def solveWithFile(dests, fileName, base=0, avoids=[]):
    gadgets = gadget.load(fileName)
    return solveWithGadget(dests, gadgets, base, avoids)

def solveWithGadget(dests, gadgets, base=0, avoids=[]):
    return solveAvoidChars(dests, gadgets, base, avoids)

def solveAvoidChars(dests, gadgets, base=0, avoids=[]):
    avoids = set(avoids)
    def cond(v):
        chars = set()
        for i in range(arch.bits() / 8):
            chars.add(v & 0xff)
            v >>= 8
        return avoids & chars == set()

    def proc(reg, dest, gadgets, canUse):
        ans = None
        if dest == 0:
            rop = toZero.find(reg, gadgets, canUse)
            if rop is not None:
                ans = util.optMin(ans, rop)

        print reg, canUse
        #use xor r1, r2
        chars = set(range(0xff)) - avoids
        xorTable = [None] * 0x100
        canConstruct = True
        for a in chars:
            for b in chars:
                xorTable[a^b] = (a, b)
        left, right = 0, 0
        for i in range(arch.bits() / 8):
            ab = xorTable[(dest >> (8*i)) & 0xff]
            if ab is None:
                canConstruct = False
                break
            left = left | ab[0] << (i * 8)
            right = right | ab[1] << (i * 8)

        if canConstruct:
            ropChainLeft = find(reg, left, gadgets, canUse, lambda x: True, lambda a, b, c, d: None)
            for r1 in canUse:
                xor = asm.xor.find(reg, r1, gadgets, canUse - set([reg, r1]))
                if xor != None:
                    ropChainRight = find(r1, right, gadgets, canUse, lambda x: True, lambda a, b, c, d: None)
                    ans = util.optMin(ans, ropChainLeft + ropChainRight + xor)

        #use reg <- somevalue; ret; (inc reg)*(value-someValue)
        tmpDest = 0
        canConstruct = True
        for i in range(arch.word()):
            byte = (dest >> (arch.bits() - 8*(i+1))) & 0xff
            filtered = filter(lambda x: x <= byte, chars)
            if len(filtered) > 0 and max(filtered) == byte:
                tmpDest = (tmpDest << 8) + byte
            elif len(filtered) > 0:
                a = max(chars)
                tmpDest = (tmpDest << (arch.bits() - 8*i)) + \
                        int((arch.word()-i)*("%x"%(a)), 16)
                break
            else:
                tmpDest -= 1
                p = lambda x: all([chr((x>>j)&0xff) for j in range(i)])
                while not p(tmpDest) and tmpDest > 0:
                    tmpDest -= 1
                if tmpDest <= 0:
                    canConstruct = False
                else:
                    a = max(chars)
                    tmpDest = (tmpDest << (arch.bits() - 8*i)) + \
                            int((arch.word()-i)*("%x"%(a)), 16)
                break

        if canConstruct:
            print hex(dest), hex(tmpDest)
            _pop = setVal.find(reg, tmpDest, gadgets, canUse)
            _inc = asm.inc.find(reg, gadgets, canUse)
            if _pop is not None and _inc is not None:
                ans = util.optMin(ans, _pop + _inc * (dest - tmpDest))

        #use reg <- 0; ret; (inc reg)*value
        zero = setVal.find(reg, 0, gadgets, canUse)
        _inc = asm.inc.find(reg, gadgets, canUse)
        if dest < 0x1000 and zero is not None and _inc is not None:
            ans = util.optMin(ans, zero + _inc * dest)

        return ans

    gadgets = list(filter(lambda g: cond(g.addr + base), gadgets))
    def uniq(xs):
        ret = [xs[0]]
        for x in xs[1:]:
            if x != ret[-1]:
                ret.append(x)
        return ret
    gadgets = sorted(gadgets, key=lambda x: len(x.changedRegs))
    gadgets = uniq(gadgets)
    # for g in gadgets:
    #     print g.toStr()

    return _solve(dests, gadgets, base, cond, proc)

def find(reg, dest, gadgets, canUse, cond, proc):
    if cond(dest):
        return setVal.find(reg, dest, gadgets, canUse)
    else:
        return proc(reg, dest, gadgets, canUse)

File: ropchain/ffi/src/common/solver.cpp
#include "solver.h"
#include "config.h"
#include "asm/xor.h"
#include "asm/inc.h"
#include "middle/middle.h"

OptROP Solver::_solve(const std::map<RegType::Reg, uint64_t>& dests, const Gadgets& gadgets,
        uint64_t base, Cond& cond, Proc& proc, const std::set<char>& avoids) {
    auto ropChains = std::map<RegType::Reg, ROPChain>();
    auto remains = RegSet();
	//Construct ROPChain by itself
    const auto allBits = Util::toBits(Util::map2Regs(dests));
    for(const auto reg : allBits) {
        const auto tmp = findROPChain(reg, dests.at(reg), gadgets,
                RegSet(reg), cond, proc, avoids);
        if(tmp.has_value()) {
            ropChains[reg] = tmp.value();
        } else {
            remains.set(reg);
        }
    }
    if(remains.none()) {
        auto rop = std::accumulate(ropChains.begin(), ropChains.end(), ROPChain(),
			[](const ROPChain& a, const auto& b){return a + b.second;});
        rop.setBaseAddr(base);
        return rop;
    }
	OptROP ans = {};
	//Brute force remains
    auto bits = Util::toBits(remains);
    std::sort(bits.begin(), bits.end());
    do {
        auto rop = ROPChain();
        auto aval = Util::allRegs();
        bool isDone = true;
        //Construct ROPChain with set of registers 'remain'
        for(RegType::Reg reg : bits) {
            const auto tmp = findROPChain(reg, dests.at(reg), gadgets, aval, cond, proc, avoids);
            if(!tmp.has_value()) {
                isDone = false;
                break;
            }
            aval.reset(reg);
            rop += tmp.value();
        }
        if(isDone) {
            ans = Util::optMin(ans, (OptROP)rop);
        }
    } while (next_permutation(bits.begin(), bits.end()));
	if(!ans.has_value()) {
		return {};
	}
	auto rop = std::accumulate(ropChains.begin(), ropChains.end(), ROPChain(),
			[](const ROPChain& a, const auto& b){return a + b.second;});
	rop = ans.value() + rop;
	rop.setBaseAddr(base);
    return (OptROP)rop;
}

//TODO move to asm/pop.cpp
//use reg <- 0; ret; (inc reg)*value
OptROP solveByInc(const RegType::Reg reg, const uint64_t dest, const Gadgets& gadgets, RegSet aval) {
    auto zero = Middle::setVal(reg, 0, gadgets, aval);
    auto inc = Inc::find(reg, gadgets, aval);
    if(dest < 0x1000 && zero.has_value() && inc.has_value()) {
        //zero + inc * dest
        auto tmp = zero.value();
        for(uint64_t i=0; i<dest; i++) {
            tmp += inc.value();
        }
        return tmp;
    }
    return {};
}

//use reg <- somevalue; ret; (inc reg)*(value-someValue)
OptROP solveByPopInc(const RegType::Reg reg, const uint64_t dest, const Gadgets& gadgets,
        const std::set<char>& avoids, const std::vector<uint8_t>& chars, RegSet aval) {
    uint64_t tmpDest = 0;
    const size_t word = Config::Arch::word();
    const uint32_t bits = Config::Arch::bits();
    for(size_t i=0; i<word; i++) {
        const uint8_t byte = (dest >> (bits - (i + 1) * 8)) & 0xff;
        std::vector<uint8_t> filtered;
        std::copy_if(chars.begin(), chars.end(),
                std::back_inserter(filtered),
                [byte](uint8_t x){return x <= byte;});
        if(filtered.size()
                && *std::max_element(filtered.begin(), filtered.end()) == byte) {
            tmpDest = (tmpDest << 8) + byte;
        } else if(filtered.size()) {
            const uint8_t a = *std::max_element(filtered.begin(), filtered.end());
            tmpDest = tmpDest << (bits - i * 8);
            for(size_t j=0; j<word-i; j++) {
                tmpDest += a << j * 8;
            }
            break;
        } else {
            tmpDest--;
            const auto p = [&i, &avoids](uint64_t t){
                for(size_t j = 0; j < i; j++) {
                    const uint8_t x = (t >> j) & 0xff;
                    if(avoids.find(x) != avoids.end()) {
                        return false;
                    }
                }
                return true;
            };
            while(!p(tmpDest) && tmpDest > 0) {
                tmpDest --;
                if(tmpDest <= 0) {
                    return {};
                } else {
                    const uint8_t a = *std::max_element(chars.begin(), chars.end());
                    tmpDest = (tmpDest << (bits - i * 8));
                    for(int j = 0; j < word - i; j++) {
                        tmpDest += a << j * 8;
                    }
                }
            }
            break;
        }
    }
    const auto pop = Middle::setVal(reg, tmpDest, gadgets, aval);
    const auto inc = Inc::find(reg, gadgets, aval);
    if(pop.has_value() && inc.has_value()) {
        //pop + inc * (dest - tmpDest)
        auto ret = pop.value();
        for(int i = 0; i < dest-tmpDest; i++) {
            ret += inc.value();
        }
        return ret;
    }
    return {};
}

//use xor r1, r2
OptROP solveByXor(const RegType::Reg reg, const uint64_t dest, const Gadgets& gadgets,
        const std::set<char>& avoids, const std::vector<uint8_t>& chars, RegSet aval) {
    OptROP rop = {};
    std::array<std::optional<std::pair<uint8_t, uint8_t>>, 0x100> xorTable;
    const std::optional<std::pair<uint8_t, uint8_t>> dummy = {};
    std::fill(xorTable.begin(), xorTable.end(), dummy);
    for(uint8_t a : chars) {
        for(uint8_t b : chars) {
            xorTable[a^b] = std::make_pair(a, b);
        }
    }
    //left ^ right == dest
    uint64_t left, right;
    left = right = 0;
    for(uint32_t i = 0; i < Config::Arch::word(); i++) {
        if(const auto ab = xorTable[(dest >> (i * 8)) & 0xff]) {
            const uint8_t l = ab.value().first;
            const uint8_t r = ab.value().second;
            left = left | l << (i * 8);
            right = right | r << (i * 8);
        } else {
            return {};
        }
    }
    const auto ropLeft = Solver::findROPChain(reg, left, gadgets, aval, {}, {}, avoids);
    if(!ropLeft.has_value()) {
        return {};
    }
    const auto allBits = Util::toBits(aval);
    for(const auto r : allBits) {
        if(const auto _xor = Xor::find(reg, r, gadgets, aval)) {
            aval.reset(reg);
            const auto ropRight =  Middle::setVal(r, right, gadgets, aval);
            aval.set(reg);
            if(ropLeft.has_value() && ropRight.has_value()) {
                const OptROP _rop = ropLeft.value() + ropRight.value() + _xor.value();
                rop = Util::optMin(rop, _rop);
                break;
            }
        }
    }
    return rop;
}

OptROP Solver::solveAvoidChars(const std::map<RegType::Reg, uint64_t>& dests, const Gadgets& gadgets,
        const uint64_t base, const std::set<char>& avoids) {
    auto cond = [&avoids](uint64_t value) {
        for(size_t i = 0; i < Config::Arch::word(); i++) {
            const char c = (char)(value & 0xff);
            if(avoids.find(c) != avoids.end()) {
                return false;
            }
            value >>= 8;
        }
        return true;
    };
    Proc proc = [](const RegType::Reg reg, const uint64_t dest,
            const Gadgets& gadgets, const std::set<char>& avoids, RegSet aval) {
        OptROP rop = {};
        if(dest == 0) {
            rop = Util::optMin(rop, Middle::toZero(reg, gadgets, aval));
        }
        //available characters
        std::vector<uint8_t> chars;
        for(uint32_t i = 0; i <= 0xff; i++) {
            if(avoids.find((uint8_t)i) == avoids.end()) {
                chars.push_back((uint8_t)i);
            }
        }
        rop = Util::optMin(rop,
                solveByXor(reg, dest, gadgets,
                    avoids, chars, aval));
        rop = Util::optMin(rop,
                solveByPopInc(reg, dest, gadgets,
                    avoids, chars, aval));
        rop = Util::optMin(rop,
            solveByInc(reg, dest, gadgets, aval));
        return rop;
    };
    auto _gadgets = Gadgets();
    std::copy_if(gadgets.begin(), gadgets.end(), std::back_inserter(_gadgets),
            [&cond, &base](auto &g) {return !g.isUseless && cond(g.addr + base);});
    return _solve(dests, Util::uniqGadgets(_gadgets), base, cond, proc, avoids);
}

OptROP Solver::solveWithGadgets(const std::map<RegType::Reg, uint64_t>& dests, const Gadgets& gadgets,
        const uint64_t base, const std::set<char>& avoids) {
    return solveAvoidChars(dests, gadgets, base, avoids);
}

OptROP Solver::solveWithFile(const std::map<RegType::Reg, uint64_t>& dests, const std::string& file,
        uint64_t base, const std::set<char>& avoids) {
    auto gadgets = Config::getGadgetLoader()(file);
    if(!gadgets.has_value()) {
        return {};
    }
    return solveWithGadgets(dests, gadgets.value(), base, avoids);
}

OptROP Solver::solveWithMap(const std::map<RegType::Reg, uint64_t>& dests,
        const std::map<uint64_t, std::string> insnStr,
        uint64_t base, const std::set<char>& avoids) {
    Gadgets gadgets;
    for(auto it : insnStr) {
        if(auto g = Util::parseGadgetString(it.first, it.second)) {
            gadgets.push_back(g.value());
        }
    }
    return solveWithGadgets(dests, gadgets, base, avoids);
}

OptROP Solver::findROPChain(const RegType::Reg reg, const uint64_t dest,
        const Gadgets& gadgets, RegSet aval, std::optional<Cond> cond,
        std::optional<Proc> proc, const std::set<char>& avoids) {
    if(!cond.has_value() || cond.value()(dest)) {
        return Middle::setVal(reg, dest, gadgets, aval);
    }
    if(proc.has_value()) {
        return proc.value()(reg, dest, gadgets, avoids, aval);
    }
    return {};
}
